Share map tile trees across synched viewports

RBBentley opened the pull request:
The map tile trees were stored with the display style - and never shared.   This made it impossible to avoid copying their tile trees when display styles were cloned.  This moved the map tile trees to the viewport where a single tile tree per viewport is used and that tile tree will be preserved and reused when the viewport is synchronized from another rather than creating a new map tile tree each time the display style is cloned.

[Bug](https://dev.azure.com/bentleycs/iModelTechnologies/_workitems/edit/592522).

pmconne commented on core/frontend/src/DisplayStyleState.ts:
```diff
@@ -655,11 +631,7 @@ export abstract class DisplayStyleState extends ElementState implements DisplayS
-    this._backgroundDrapeMap.setBaseLayerSettings(this.settings.mapImagery.backgroundBase);
-    this._backgroundDrapeMap.setLayerSettings(this.settings.mapImagery.backgroundLayers);
-    this._overlayMap.setLayerSettings(this.settings.mapImagery.overlayLayers);
+    this.onMapImageryChanged.raiseEvent(this.settings.mapImagery);
```
DisplayStyleSettings has an onMapImageryChanged event that is raised by synchMapImagery. Why does DisplayStyleState also need one?

RBBentley replied:
No not needed - I removed the unnecessary DisplayStyleState.onMapImageryChanged as well as DisplayStyleState.onMapSettingsChanged.

pmconne commented on core/frontend/src/Viewport.ts:
```diff
@@ -1126,6 +1143,11 @@ export abstract class Viewport implements IDisposable {
     if (this._view)
       this._view.detachFromViewport();
+
+    if (this._mapTiledGraphicsProvider) {
```
Shouldn't this happen in Viewport.detachFromDisplayStyle, not detachFromView? And then Viewport.registerDisplayStyleListeners would create a new MapTiledGraphicsProvider if we are changing the display style?

RBBentley replied:
Fixed.

pmconne reviewed (commented):
apparently i have to leave a comment to leave several other comments.

mdastous-bentley commented:
map-layers widget depends on events to auto-refresh when something change in the view... I need to revisit a couple of workflows

mdastous-bentley reviewed (approved):
Tested map-layers widget, events are correctly fired and UI reacts as expected.

pmconne commented:
Weird things happen if you have two synchronized viewports displaying the map and you change the map type in one of them.
Sometimes, both viewports reload the map immediately and display correctly.
Sometimes, one of the viewports does not update immediately. If you zoom out, it will update with tiles of the new map type, but it will continue to also display any tiles it previously loaded for the previous map type.
Once, it crashed on an assertion:
```
Error: Assert: Programmer Error
at Object.assert (http://localhost:3000/static/js/main.chunk.js:12554:15)
at MapTileTree.getLayerTransparency (http://localhost:3000/static/js/main.chunk.js:120875:26)
at MapTile.computeDrapeTexture (http://localhost:3000/static/js/main.chunk.js:120175:176)
at MapTile.getDrapeTextures (http://localhost:3000/static/js/main.chunk.js:120153:49)
at MapTile.produceGraphics (http://localhost:3000/static/js/main.chunk.js:120025:31)
at RealityTileDrawArgs.getTileGraphics (http://localhost:3000/static/js/main.chunk.js:115011:21)
at MapTileTree.draw (http://localhost:3000/static/js/main.chunk.js:113516:35)
at MapTileTreeReference.addToScene (http://localhost:3000/static/js/main.chunk.js:121181:18)
at http://localhost:3000/static/js/main.chunk.js:71064:57
at http://localhost:3000/static/js/main.chunk.js:74023:78
at MapTiledGraphicsProvider.forEachTileTreeRef (http://localhost:3000/static/js/main.chunk.js:121350:13)
at ScreenViewport.forEachMapTreeRef (http://localhost:3000/static/js/main.chunk.js:74023:44)
at SpatialViewState.createScene (http://localhost:3000/static/js/main.chunk.js:71064:26)
at SpatialViewState.createScene (http://localhost:3000/static/js/main.chunk.js:67960:15)
at ScreenViewport.renderFrame (http://localhost:3000/static/js/main.chunk.js:74738:22)
at ViewManager.renderLoop (http://localhost:3000/static/js/main.chunk.js:70382:16)
at eventLoop (http://localhost:3000/static/js/main.chunk.js:62592:35)
```

RBBentley commented:
> Weird things happen if you have two synchronized viewports displaying the map and you change the map type in one of them.
> Sometimes, both viewports reload the map immediately and display correctly.
> Sometimes, one of the viewports does not update immediately. If you zoom out, it will update with tiles of the new map type, but it will continue to also display any tiles it previously loaded for the previous map type.
> Once, it crashed on an assertion:
> 
> ```
> Error: Assert: Programmer Error
> at Object.assert (http://localhost:3000/static/js/main.chunk.js:12554:15)
> at MapTileTree.getLayerTransparency (http://localhost:3000/static/js/main.chunk.js:120875:26)
> at MapTile.computeDrapeTexture (http://localhost:3000/static/js/main.chunk.js:120175:176)
> at MapTile.getDrapeTextures (http://localhost:3000/static/js/main.chunk.js:120153:49)
> at MapTile.produceGraphics (http://localhost:3000/static/js/main.chunk.js:120025:31)
> at RealityTileDrawArgs.getTileGraphics (http://localhost:3000/static/js/main.chunk.js:115011:21)
> at MapTileTree.draw (http://localhost:3000/static/js/main.chunk.js:113516:35)
> at MapTileTreeReference.addToScene (http://localhost:3000/static/js/main.chunk.js:121181:18)
> at http://localhost:3000/static/js/main.chunk.js:71064:57
> at http://localhost:3000/static/js/main.chunk.js:74023:78
> at MapTiledGraphicsProvider.forEachTileTreeRef (http://localhost:3000/static/js/main.chunk.js:121350:13)
> at ScreenViewport.forEachMapTreeRef (http://localhost:3000/static/js/main.chunk.js:74023:44)
> at SpatialViewState.createScene (http://localhost:3000/static/js/main.chunk.js:71064:26)
> at SpatialViewState.createScene (http://localhost:3000/static/js/main.chunk.js:67960:15)
> at ScreenViewport.renderFrame (http://localhost:3000/static/js/main.chunk.js:74738:22)
> at ViewManager.renderLoop (http://localhost:3000/static/js/main.chunk.js:70382:16)
> at eventLoop (http://localhost:3000/static/js/main.chunk.js:62592:35)
> ```

Fixed... although it is not pretty.   I have code to handle changing the imagery layers without reloading the tiles when the layers are changed through the layer API but synching the view was bypassing these - I added a method on the graphics provider to handle this case explicitly.

pmconne commented on core/frontend/src/ViewState.ts:
```diff
@@ -256,7 +256,7 @@ export abstract class ViewState extends ElementState {
   }
 
-  /** Returns true if all [[TileTree]]s required by this view have been loaded. */
+  /** Returns true if all [[TileTree]]s required by this view have been loaded.  Note that the map tile trees associated to the viewport rather than the view. */
```
Can you add a link to `Viewport.areAllTileTreesLoaded` here and a note to NextVersion.md describing the change in behavior?

RBBentley replied:
Yes - I wasn't sure of this one.  Fixed.

pmconne commented on core/frontend/src/SheetViewState.ts:
```diff
@@ -869,7 +869,7 @@ class RasterAttachment {
 
   public get areAllTileTreesLoaded() {
-    return this._viewport?.view.areAllTileTreesLoaded ?? true;
+    return this._viewport?.view?.areAllTileTreesLoaded ?? true;
```
Shouldn't this check `this._viewport.areAllTileTreesLoaded`? The background map might be turned on in the attachment.

RBBentley replied:
Yes - I wasn't sure of this one.   Fixed.

pmconne reviewed (changes requested):
.

pmconne commented:
Failing integration test looks potentially related to these changes.
```
  1) Section Drawings (#integration)
       displays the 3d tiles in the 2d view if so specified:

      AssertionError: expected 0 to be at least 1
      + expected - actual

      -0
      +1
      
      at home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:184:26
      at home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:168:9
      at Object.testOnScreenViewport (home/vsts/work/1/s/full-stack-tests/core/src/frontend/TestViewport.ts:336:5)
      at test (home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:166:7)
      at Context.<anonymous> (home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:182:5)
```

RBBentley commented:
> Failing integration test looks potentially related to these changes.
> 
> ```
>   1) Section Drawings (#integration)
>        displays the 3d tiles in the 2d view if so specified:
> 
>       AssertionError: expected 0 to be at least 1
>       + expected - actual
> 
>       -0
>       +1
>       
>       at home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:184:26
>       at home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:168:9
>       at Object.testOnScreenViewport (home/vsts/work/1/s/full-stack-tests/core/src/frontend/TestViewport.ts:336:5)
>       at test (home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:166:7)
>       at Context.<anonymous> (home/vsts/work/1/s/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:182:5)
> ```

Investigating now - but am not able to reproduce locally, can you reproduce this?

Ray

pmconne commented:
> Investigating now - but am not able to reproduce locally, can you reproduce this?

Yes, on first try. TestViewport uses areAllTileTreesLoaded to make sure test waits until all tile trees load before continuing. I see you did update it to use Viewport instead of ViewState.

```
[j2  7:10:09.19] D:\j2\s\imodeljs\full-stack-tests\core>npm run test:integration:electron

> core-full-stack-tests@ test:integration:electron D:\j2\s\imodeljs\full-stack-tests\core
> certa -r electron --grep "#integration"


(node:10112) electron: The default of contextIsolation is deprecated and will be changing from false to true in a future release of Electron.  See https://github.com/electron/electron/issues/23506 for more information


  Section Drawings (#integration)

DevTools listening on ws://127.0.0.1:60791/devtools/browser/e8064d4d-9e90-41a8-856c-f8cddecf8b78
[0504/071028.885:INFO:CONSOLE(357)] "isWebAuthnPlatformAuthenticatorAvailable - false", source: https://qa-imsoidc.bentley.com/as/moO3a/resume/as/authorization.ping (357)
    √ instantiates a SectionDrawingModelState
    √ loads section drawing info for view (89ms)
    √ updates section drawing info when viewed model changes
    √ clones section drawing info
    √ preserves section drawing info when round-tripped through JSON
    √ clones attachment info when view is cloned
    √ only allocates attachment if attachment is to be displayed (79ms)
    √ allocates attachment when attached to viewport and disposes of it when detached from viewport
    1) displays the 3d tiles in the 2d view if so specified


  8 passing (14s)
  1 failing

  1) Section Drawings (#integration)
       displays the 3d tiles in the 2d view if so specified:

      AssertionError: expected 0 to be at least 1
      + expected - actual

      -0
      +1

      at D:/j2/s/imodeljs/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:184:26
      at D:/j2/s/imodeljs/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:168:9
      at Object.testOnScreenViewport (D:/j2/s/imodeljs/full-stack-tests/core/src/frontend/TestViewport.ts:336:5)
      at test (D:/j2/s/imodeljs/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:166:7)
      at Context.<anonymous> (D:/j2/s/imodeljs/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts:182:5)



npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! core-full-stack-tests@ test:integration:electron: `certa -r electron --grep "#integration"`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the core-full-stack-tests@ test:integration:electron script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\paul.connelly\AppData\Roaming\npm-cache\_logs\2021-05-04T11_10_36_607Z-debug.log

[j2  7:10:36.62] D:\j2\s\imodeljs\full-stack-tests\core>git diff -w
diff --git a/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts b/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts
index b1a3a3e66d..d6f2c3c473 100644
--- a/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts
+++ b/full-stack-tests/core/src/frontend/hub/SectionDrawing.test.ts
@@ -8,7 +8,7 @@ import { TestUsers } from "@bentley/oidc-signin-tool/lib/TestUsers";
 import { testOnScreenViewport, TestViewport } from "../TestViewport";
 import { TestUtility } from "./TestUtility";

-describe("Section Drawings (#integration)", () => {
+describe.only("Section Drawings (#integration)", () => {
   let imodel: IModelConnection;

   before(async () => {
```

pmconne commented on core/frontend/src/Viewport.ts:
```diff
@@ -1353,17 +1379,37 @@ export abstract class Viewport implements IDisposable {
+  /**
+   * Returns true if all [[TileTree]]s required by this viewport have been loaded.
+   */
+  public get areAllTileTreesLoaded(): boolean {
```
This needs to invoke `this.view.areAllTileTreesLoaded`. It should then check all tile tree refs owned by the viewport (and not those owned by the ViewState).
See specializations of `areAllTileTreesLoaded` by SheetViewState and DrawingViewState.

pmconne replied:
Pushed this change.
In a separate PR I think we should replace all the `forEachXXXRef` functions with `Iterable<TileTreeReference>`s - they would be less ugly and allow you to `break`.
